File: packages/virage-core/src/quality/runner.ts
/**
 * QualityRunner — orchestrates all 8 pipeline component metric collections.
 *
 * Loads config, samples chunks from VirageDb, runs each component's metrics
 * in dependency order, aggregates scores, and returns a QualityReport.
 * Optional components (5-8) are skipped gracefully when not configured.
 */

import type {
  QualityReport,
  QualityRunnerOptions,
  ComponentResult,
  MustPassGate,
  MetricResult,
  RagBenchSummary,
} from "./interfaces.js";
import { loadRagBenchDataset, RagBenchEvaluator } from "../eval/ragbench.js";
import {
  aggregateComponent,
  aggregateOverall,
  computeStatus,
} from "./scoring.js";
import { computeChunkingMetrics } from "./metrics/chunking.js";
import { computeMetadataMetrics } from "./metrics/metadata.js";
import { computeDenseInputMetrics } from "./metrics/dense-input.js";
import { computeDenseEmbeddingMetrics } from "./metrics/dense-embedding.js";
import { computeSparseInputMetrics } from "./metrics/sparse-input.js";
import { computeLexicalRetrievalMetrics } from "./metrics/lexical-retrieval.js";
import { computeRerankerInputMetrics } from "./metrics/reranker-input.js";
import { computeRerankerMetrics } from "./metrics/reranker.js";
import { loadConfig } from "../config-loader.js";
import { VirageDb } from "../core/virage-db.js";
import { defaultVirageDb } from "../core/virage-defaults.js";
import type { Chunk } from "../interfaces/index.js";

// Component total weights used for overall aggregation
const COMPONENT_WEIGHTS: Record<string, number> = {
  chunking: 2.0 + 2.0 + 0.5 + 1.0, // sum of metric weights
  metadata: 1.0 + 1.0 + 1.0 + 1.0 + 0.5,
  denseInput: 1.0 + 1.0,
  denseEmbedding: 3.0 + 1.0 + 1.0 + 0.5 + 1.0,
  sparseInput: 1.0 + 0.5,
  lexicalRetrieval: 1.5,
  rerankerInput: 1.0 + 1.0 + 0.5 + 1.0,
  reranker: 2.5 + 1.0 + 0.5,
};

function collectMustPassGates(components: ComponentResult[]): MustPassGate[] {
  const gates: MustPassGate[] = [];
  for (const comp of components) {
    for (const m of comp.metrics) {
      if (m.mustPass && !m.skipped && m.mustPassPassed !== undefined) {
        gates.push({
          metricName: m.name,
          threshold: m.mustPassThreshold ?? 0,
          value: m.rawValue,
          passed: m.mustPassPassed,
        });
      }
    }
  }
  return gates;
}

function metricsToComponent(
  id: string,
  label: string,
  metrics: MetricResult[],
  componentWeight: number,
  skipped = false,
  skipReason?: string,
): ComponentResult {
  const score = aggregateComponent(metrics);
  return {
    id: id as ComponentResult["id"],
    label,
    score,
    weight: componentWeight,
    skipped,
    skipReason,
    metrics,
  };
}

function sampleArray<T>(arr: T[], n: number): T[] {
  if (arr.length <= n) return arr;
  const step = arr.length / n;
  return Array.from({ length: n }, (_, i) => arr[Math.floor(i * step)]);
}

export async function runQualityAssessment(
  opts: QualityRunnerOptions,
): Promise<QualityReport> {
  const t0 = Date.now();
  const {
    configFile,
    sampleSize,
    topK,
    failFast,
    ragBenchPath,
    thresholdOverrides = {},
    weightOverrides = {},
  } = opts;

  const cfg = await loadConfig(configFile);

  // Merge config.quality thresholds/weights into options (CLI flags take precedence)
  const qualityCfg = cfg.quality;
  const resolvedThresholds = {
    ...qualityCfg?.thresholds,
    ...thresholdOverrides,
  };
  const resolvedWeights = { ...qualityCfg?.weights, ...weightOverrides };

  await cfg.vectorStore.initialize();

  const db = new VirageDb(defaultVirageDb());
  let allChunks: Chunk[];
  try {
    allChunks = db.getAllChunks();
  } finally {
    db.close();
  }

  if (allChunks.length === 0) {
    throw new Error("No chunks found in virage.db — run `virage index` first.");
  }

  const sample = sampleArray(allChunks, sampleSize);
  const chunkIdSet = new Set(allChunks.map((c) => c.denseTextHash));

  // Dense search function wrapper
  const searchFn = async (query: string, k: number) => {
    const embed = await cfg.embedder.embed(query);
    const results = await cfg.vectorStore.search(embed, k);
    return results.map((r) => ({ id: r.id }));
  };

  // Lexical (FTS) search — attempt hybrid with alpha=0 (pure BM25)
  let ftsSearchFn:
    | ((q: string, k: number) => Promise<Array<{ id: string }>>)
    | null = null;
  try {
    const testEmbed = await cfg.embedder.embed("test");
    await cfg.vectorStore.search(testEmbed, 1, undefined, {
      hybrid: true,
      hybridAlpha: 0,
      queryText: "test",
    });
    ftsSearchFn = async (query: string, k: number) => {
      const embed = await cfg.embedder.embed(query);
      const results = await cfg.vectorStore.search(embed, k, undefined, {
        hybrid: true,
        hybridAlpha: 0,
        queryText: query,
      });
      return results.map((r) => ({ id: r.id }));
    };
  } catch {
    // Vector store doesn't support hybrid/FTS — lexical component will be skipped
  }

  // Reranker availability
  const rerankerAvailable = cfg.search?.reranker != null;

  const components: ComponentResult[] = [];

  // ─── Component 1: Chunking ────────────────────────────────────────────────
  const chunkingMetrics = await computeChunkingMetrics(
    {
      chunks: sample.map((c) => ({
        denseText: c.denseText,
        sparseText: c.sparseText,
        metadata: {
          astNodeCount: undefined,
          astNodeCountInBounds: undefined,
          tokenCount: c.metadata?.estimatedTokens,
        },
      })),
      embedFn: (t) => cfg.embedder.embed(t),
      tokenRangeMin: 50,
      tokenRangeMax: 512,
    },
    Math.min(50, sampleSize),
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "chunking",
      "Chunking",
      chunkingMetrics,
      COMPONENT_WEIGHTS.chunking,
    ),
  );

  if (
    failFast &&
    components.some((c) => c.metrics.some((m) => m.mustPassPassed === false))
  ) {
    return buildReport(t0, components, sampleSize, topK, configFile, undefined);
  }

  // ─── Component 2: Metadata ────────────────────────────────────────────────
  const metadataMetrics = computeMetadataMetrics(
    {
      chunks: sample.map((c) => ({
        metadata: {
          id: c.denseTextHash,
          breadcrumb: Array.isArray(c.metadata?.breadcrumb)
            ? c.metadata.breadcrumb.join(" > ")
            : undefined,
          fqn: c.metadata?.fqn,
          imports: c.metadata?.imports,
          resolvedImports: undefined,
          totalImports: c.metadata?.imports?.length,
          prevId: c.metadata?.siblingPrev,
          nextId: c.metadata?.siblingNext,
          sourceFile: c.metadata?.sourceFile,
          isCode: !!c.metadata?.codeLanguage,
        },
        denseText: c.denseText,
      })),
      chunkIdSet,
    },
    resolvedThresholds["importResolution"] ?? 0.7,
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "metadata",
      "Metadata Extraction",
      metadataMetrics,
      COMPONENT_WEIGHTS.metadata,
    ),
  );

  if (
    failFast &&
    components.some((c) => c.metrics.some((m) => m.mustPassPassed === false))
  ) {
    return buildReport(t0, components, sampleSize, topK, configFile, undefined);
  }

  // ─── Component 3: Dense Input ─────────────────────────────────────────────
  const denseInputMetrics = await computeDenseInputMetrics(
    {
      chunks: sample.map((c) => ({
        denseText: c.denseText,
        sparseText: c.sparseText,
      })),
      embedFn: (t) => cfg.embedder.embed(t),
    },
    Math.min(30, sampleSize),
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "denseInput",
      "Dense Input Prep",
      denseInputMetrics,
      COMPONENT_WEIGHTS.denseInput,
    ),
  );

  // ─── Component 4: Dense Embedding ─────────────────────────────────────────
  const denseEmbeddingMetrics = await computeDenseEmbeddingMetrics(
    {
      chunks: sample.map((c) => ({
        id: c.denseTextHash,
        denseText: c.denseText,
      })),
      searchFn,
      embedFn: (t) => cfg.embedder.embed(t),
      topK,
    },
    Math.min(sampleSize, 200),
    resolvedThresholds["selfRecall"] ?? 0.8,
    resolvedThresholds["outlierFraction"] ?? 0.05,
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "denseEmbedding",
      "Dense Embedding",
      denseEmbeddingMetrics,
      COMPONENT_WEIGHTS.denseEmbedding,
    ),
  );

  if (
    failFast &&
    components.some((c) => c.metrics.some((m) => m.mustPassPassed === false))
  ) {
    return buildReport(t0, components, sampleSize, topK, configFile, undefined);
  }

  // ─── Component 5: Sparse Input (optional) ─────────────────────────────────
  const sparseChunks = sample.filter((c) => !!c.sparseText);
  const sparseInputMetrics = computeSparseInputMetrics(
    { chunks: sparseChunks.map((c) => ({ sparseText: c.sparseText })) },
    resolvedWeights,
  );
  const sparseSkipped = sparseChunks.length === 0;
  components.push(
    metricsToComponent(
      "sparseInput",
      "Sparse Input Prep",
      sparseInputMetrics,
      COMPONENT_WEIGHTS.sparseInput,
      sparseSkipped,
      sparseSkipped ? "No sparseText in sampled chunks" : undefined,
    ),
  );

  // ─── Component 6: Lexical Retrieval (optional) ────────────────────────────
  const lexicalMetrics = await computeLexicalRetrievalMetrics(
    {
      chunks: sample.map((c) => ({
        id: c.denseTextHash,
        denseText: c.denseText,
      })),
      ftsSearchFn,
      topK,
    },
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "lexicalRetrieval",
      "Lexical Retrieval",
      lexicalMetrics,
      COMPONENT_WEIGHTS.lexicalRetrieval,
    ),
  );

  // ─── Component 7: Reranker Input (optional) ───────────────────────────────
  const rerankerInputMetrics = computeRerankerInputMetrics(
    { samples: [], rerankerAvailable },
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "rerankerInput",
      "Reranker Input",
      rerankerInputMetrics,
      COMPONENT_WEIGHTS.rerankerInput,
      !rerankerAvailable,
      !rerankerAvailable ? "No reranker configured" : undefined,
    ),
  );

  // ─── Component 8: Reranker (optional) ────────────────────────────────────
  const rerankerMetrics = computeRerankerMetrics(
    {
      rerankerMrr: null,
      baselineMrr: null,
      rerankerScores: [],
      rerankerAvailable,
    },
    resolvedWeights,
  );
  components.push(
    metricsToComponent(
      "reranker",
      "Reranker",
      rerankerMetrics,
      COMPONENT_WEIGHTS.reranker,
      !rerankerAvailable,
      !rerankerAvailable ? "No reranker configured" : undefined,
    ),
  );

  // ─── RAGBench evaluation (optional) ─────────────────────────────────────
  let ragBench: RagBenchSummary | undefined;
  if (ragBenchPath) {
    const dataset = await loadRagBenchDataset(ragBenchPath);
    const evaluator = new RagBenchEvaluator(cfg.vectorStore, cfg.embedder);
    const result = await evaluator.evaluate(dataset, topK);
    ragBench = result;
  }

  return buildReport(t0, components, sampleSize, topK, configFile, ragBench);
}

function buildReport(
  t0: number,
  components: ComponentResult[],
  sampleSize: number,
  topK: number,
  configFile: string,
  ragBench: RagBenchSummary | undefined,
): QualityReport {
  const mustPassGates = collectMustPassGates(components);
  const overallScore = aggregateOverall(components);
  const status = computeStatus(overallScore, mustPassGates);

  return {
    timestamp: new Date().toISOString(),
    overallScore,
    status,
    mustPassGates,
    components,
    ...(ragBench ? { ragBench } : {}),
    sampleSize,
    topK,
    configFile,
    durationMs: Date.now() - t0,
  };
}

File: crates/virage-engine/src/quality/runner.rs
//! QualityRunner — orchestrates all 8 pipeline component metric collections.
//!
//! Loads the vector store, samples chunks, runs each component's metrics in dependency
//! order, aggregates scores, and returns a `QualityReport`. Optional components (5-8) are
//! skipped gracefully when not configured. Ported from `dist/quality/runner.js` — IR-038.

use std::collections::HashSet;
use std::sync::Mutex;

use crate::config::VirageConfigJson;
use crate::embedders::Embedder;
use crate::stores::{SearchOptions, VectorStore};

use super::metrics::{
    chunking::compute_chunking_metrics, dense_embedding::compute_dense_embedding_metrics,
    dense_input::compute_dense_input_metrics, lexical_retrieval::compute_lexical_retrieval_metrics,
    metadata::compute_metadata_metrics, reranker::compute_reranker_metrics,
    reranker_input::compute_reranker_input_metrics, sparse_input::compute_sparse_input_metrics,
};
use super::scoring::{aggregate_overall, compute_status};
use super::{ComponentId, ComponentResult, MustPassGate, QualityChunk, QualityReport};

// Component total weights, used for overall aggregation. Sum of each component's metric
// weights, matching the JS predecessor's `COMPONENT_WEIGHTS`.
const WEIGHT_CHUNKING: f64 = 2.0 + 2.0 + 0.5 + 1.0;
const WEIGHT_METADATA: f64 = 1.0 + 1.0 + 1.0 + 1.0 + 0.5;
const WEIGHT_DENSE_INPUT: f64 = 1.0 + 1.0;
const WEIGHT_DENSE_EMBEDDING: f64 = 3.0 + 1.0 + 1.0 + 0.5 + 1.0;
const WEIGHT_SPARSE_INPUT: f64 = 1.0 + 0.5;
const WEIGHT_LEXICAL_RETRIEVAL: f64 = 1.5;
const WEIGHT_RERANKER_INPUT: f64 = 1.0 + 1.0 + 0.5 + 1.0;
const WEIGHT_RERANKER: f64 = 2.5 + 1.0 + 0.5;

pub struct QualityRunOptions {
    pub config_file: String,
    pub sample_size: usize,
    pub top_k: usize,
}

fn sample_evenly(chunks: Vec<QualityChunk>, n: usize) -> Vec<QualityChunk> {
    if chunks.len() <= n {
        return chunks;
    }
    let step = chunks.len() as f64 / n as f64;
    (0..n)
        .map(|i| chunks[(i as f64 * step) as usize].clone())
        .collect()
}

fn collect_must_pass_gates(components: &[ComponentResult]) -> Vec<MustPassGate> {
    let mut gates = Vec::new();
    for comp in components {
        for m in &comp.metrics {
            if m.must_pass && !m.skipped {
                if let Some(passed) = m.must_pass_passed {
                    gates.push(MustPassGate {
                        metric_name: m.name.clone(),
                        threshold: m.must_pass_threshold.unwrap_or(0.0),
                        value: m.raw_value,
                        passed,
                    });
                }
            }
        }
    }
    gates
}

async fn compute_recall(
    store: &dyn VectorStore,
    embedder: &Mutex<dyn Embedder + Send>,
    chunks: &[QualityChunk],
    top_k: usize,
    opts: SearchOptions,
) -> anyhow::Result<f64> {
    if chunks.is_empty() {
        return Ok(0.0);
    }
    let mut hits = 0;
    for chunk in chunks {
        let query: String = chunk.dense_text.chars().take(80).collect();
        let vec = {
            let mut guard = embedder
                .lock()
                .map_err(|_| anyhow::anyhow!("embedder lock poisoned"))?;
            guard
                .embed_batch(std::slice::from_ref(&query))
                .map_err(|e| anyhow::anyhow!("embed error: {e}"))?
        };
        let search_opts = SearchOptions {
            hybrid: opts.hybrid,
            hybrid_alpha: opts.hybrid_alpha,
            query_text: if opts.hybrid {
                Some(query.clone())
            } else {
                None
            },
            filter: None,
            tag_filter: None,
        };
        let results = store.search(&vec, top_k, search_opts).await?;
        if results.iter().any(|r| r.id == chunk.id) {
            hits += 1;
        }
    }
    Ok(hits as f64 / chunks.len() as f64)
}

pub async fn run_quality_assessment(
    cfg: &VirageConfigJson,
    dims: usize,
    opts: &QualityRunOptions,
) -> anyhow::Result<QualityReport> {
    let t0 = std::time::Instant::now();

    let thresholds = cfg.quality.as_ref();
    let import_threshold = thresholds
        .map(|q| q.import_resolution_min())
        .unwrap_or(crate::config::QualityThresholds::DEFAULT_IMPORT_RESOLUTION_MIN);
    let self_recall_threshold = thresholds
        .map(|q| q.self_recall_min())
        .unwrap_or(crate::config::QualityThresholds::DEFAULT_SELF_RECALL_MIN);
    let outlier_threshold = thresholds
        .map(|q| q.outlier_fraction_max())
        .unwrap_or(crate::config::QualityThresholds::DEFAULT_OUTLIER_FRACTION_MAX);

    let embedder = crate::config::resolve::resolve_embedder(&cfg.providers.embedder)?;
    let store = crate::config::resolve::resolve_store(&cfg.providers.vector_store, dims)?;
    store.initialize().await?;

    let Some(all_results) = store.list_all().await? else {
        anyhow::bail!(
            "Vector store does not support listing all chunks — cannot run quality assessment."
        );
    };
    if all_results.is_empty() {
        anyhow::bail!("No chunks found in vector store — run `virage index` first.");
    }
    let all_chunks: Vec<QualityChunk> = all_results
        .into_iter()
        .map(|r| QualityChunk {
            id: r.id,
            dense_text: r.dense_text,
            sparse_text: r.sparse_text,
            metadata: r.metadata,
            source_file: r.source_file,
        })
        .collect();
    let chunk_id_set: HashSet<String> = all_chunks.iter().map(|c| c.id.clone()).collect();
    let sample = sample_evenly(all_chunks, opts.sample_size);

    // Reranker availability.
    let reranker_available = cfg.providers.reranker.is_some();

    let mut components = Vec::with_capacity(8);

    // ─── Component 1: Chunking ────────────────────────────────────────────────
    let chunking_metrics = compute_chunking_metrics(
        &sample,
        embedder.as_ref(),
        opts.sample_size.min(50),
        50.0,
        512.0,
    )?;
    components.push(ComponentResult::new(
        ComponentId::Chunking,
        "Chunking",
        chunking_metrics,
        WEIGHT_CHUNKING,
        false,
        None,
    ));

    // ─── Component 2: Metadata ────────────────────────────────────────────────
    let metadata_metrics = compute_metadata_metrics(&sample, &chunk_id_set, import_threshold);
    components.push(ComponentResult::new(
        ComponentId::Metadata,
        "Metadata Extraction",
        metadata_metrics,
        WEIGHT_METADATA,
        false,
        None,
    ));

    // ─── Component 3: Dense Input ─────────────────────────────────────────────
    let dense_input_metrics =
        compute_dense_input_metrics(&sample, embedder.as_ref(), opts.sample_size.min(30))?;
    components.push(ComponentResult::new(
        ComponentId::DenseInput,
        "Dense Input Prep",
        dense_input_metrics,
        WEIGHT_DENSE_INPUT,
        false,
        None,
    ));

    // ─── Component 4: Dense Embedding ─────────────────────────────────────────
    let dense_sample: Vec<QualityChunk> = sample
        .iter()
        .take(opts.sample_size.min(200))
        .cloned()
        .collect();
    let self_recall = compute_recall(
        store.as_ref(),
        embedder.as_ref(),
        &dense_sample,
        opts.top_k,
        SearchOptions::default(),
    )
    .await?;
    let vectors: Vec<Vec<f32>> = {
        let mut guard = embedder
            .lock()
            .map_err(|_| anyhow::anyhow!("embedder lock poisoned"))?;
        let d = guard.dimensions();
        let texts: Vec<String> = dense_sample.iter().map(|c| c.dense_text.clone()).collect();
        if texts.is_empty() {
            Vec::new()
        } else {
            let flat = guard
                .embed_batch(&texts)
                .map_err(|e| anyhow::anyhow!("embed error: {e}"))?;
            flat.chunks(d).map(|c| c.to_vec()).collect()
        }
    };
    let dense_embedding_metrics = compute_dense_embedding_metrics(
        self_recall,
        &vectors,
        self_recall_threshold,
        outlier_threshold,
    );
    components.push(ComponentResult::new(
        ComponentId::DenseEmbedding,
        "Dense Embedding",
        dense_embedding_metrics,
        WEIGHT_DENSE_EMBEDDING,
        false,
        None,
    ));

    // ─── Component 5: Sparse Input (optional) ─────────────────────────────────
    let sparse_texts: Vec<String> = sample
        .iter()
        .filter(|c| !c.sparse_text.is_empty())
        .map(|c| c.sparse_text.clone())
        .collect();
    let sparse_skipped = sparse_texts.is_empty();
    let sparse_input_metrics = compute_sparse_input_metrics(&sparse_texts, None);
    components.push(ComponentResult::new(
        ComponentId::SparseInput,
        "Sparse Input Prep",
        sparse_input_metrics,
        WEIGHT_SPARSE_INPUT,
        sparse_skipped,
        sparse_skipped.then(|| "No sparseText in sampled chunks".to_string()),
    ));

    // ─── Component 6: Lexical Retrieval (optional) ────────────────────────────
    let fts_probe = store
        .search(
            &vec![0.0_f32; dims],
            1,
            SearchOptions {
                hybrid: true,
                hybrid_alpha: 0.0,
                query_text: Some("test".to_string()),
                filter: None,
                tag_filter: None,
            },
        )
        .await;
    let lexical_metrics = if fts_probe.is_err() {
        compute_lexical_retrieval_metrics(
            None,
            "FTS/BM25 search not available on this vector store",
        )
    } else if sample.is_empty() {
        compute_lexical_retrieval_metrics(None, "No chunks in sample")
    } else {
        let lexical_recall = compute_recall(
            store.as_ref(),
            embedder.as_ref(),
            &sample,
            opts.top_k,
            SearchOptions {
                hybrid: true,
                hybrid_alpha: 0.0,
                query_text: None,
                filter: None,
                tag_filter: None,
            },
        )
        .await?;
        compute_lexical_retrieval_metrics(Some(lexical_recall), "")
    };
    components.push(ComponentResult::new(
        ComponentId::LexicalRetrieval,
        "Lexical Retrieval",
        lexical_metrics,
        WEIGHT_LEXICAL_RETRIEVAL,
        false,
        None,
    ));

    // ─── Component 7: Reranker Input (optional) ───────────────────────────────
    let reranker_input_metrics = compute_reranker_input_metrics(reranker_available);
    components.push(ComponentResult::new(
        ComponentId::RerankerInput,
        "Reranker Input",
        reranker_input_metrics,
        WEIGHT_RERANKER_INPUT,
        !reranker_available,
        (!reranker_available).then(|| "No reranker configured".to_string()),
    ));

    // ─── Component 8: Reranker (optional) ─────────────────────────────────────
    let reranker_metrics = compute_reranker_metrics(reranker_available);
    components.push(ComponentResult::new(
        ComponentId::Reranker,
        "Reranker",
        reranker_metrics,
        WEIGHT_RERANKER,
        !reranker_available,
        (!reranker_available).then(|| "No reranker configured".to_string()),
    ));

    let must_pass_gates = collect_must_pass_gates(&components);
    let overall_score = aggregate_overall(&components);
    let status = compute_status(overall_score, &must_pass_gates);

    Ok(QualityReport {
        timestamp: crate::history::timestamp_now_iso(),
        overall_score,
        status,
        must_pass_gates,
        components,
        sample_size: opts.sample_size,
        top_k: opts.top_k,
        config_file: opts.config_file.clone(),
        duration_ms: t0.elapsed().as_millis(),
    })
}
